### ISSUE TITLE
loss 在0.7500上下跳动

### ISSUE THREAD

**Issue by milk-bottle-liyu** [tyandzx/caffe#5]
您好，我由于项目需要，临时了解了一些尺度学习的知识，所以感觉用起来有很多地方并不了解，所以想请教一下。

1. 我看前面的issue有说，您采用的是三角的loss，不是很懂这个，我还以为是l2的loss，请问下这个loss的全称或者公式？

2.我们一开始觉得可能和别的任务一样，深层网络较好，所以用了Res100和Res50去训练，但是loss一直停留在0.75（margin 0.25），不是波动，而是保持在

3. 我们最后采用了ZF-net结构
 `name:"ZF car detect"
layer {
  name: "data"
  type: "ImageData"
  top: "data"
  top: "label"
  transform_param {
    mean_value: 127.5
    mean_value: 127.5
    mean_value: 127.5
    scale: 0.0078125
#    mirror: true
  }
  image_data_param {
    new_height: 256
    new_width: 256 
    source: "imagesets/train.txt"
    batch_size: 32
    shuffle: false
  }
}

layer {
	name: "conv1"
	type: "Convolution"
	bottom: "data"
	top: "conv1"
	param { lr_mult: 1.0 }
	param { lr_mult: 2.0 }
	convolution_param {
		num_output: 96
		kernel_size: 7
		stride: 2
		weight_filler {
			type: "xavier"
		}
		bias_filler {
			type: "constant"
			value: 0
		}
	}
}

layer {
	name: "relu1"
	type: "ReLU"
	bottom: "conv1"
	top: "conv1"
}

layer {
	name: "pool1"
	type: "Pooling"
	bottom: "conv1"
	top: "pool1"
	pooling_param {
		kernel_size: 3
		stride: 2
		pool: MAX
	}
}

layer {
	name: "conv2"
	type: "Convolution"
	bottom: "pool1"
	top: "conv2"
	param { lr_mult: 1.0 }
	param { lr_mult: 2.0 }
	convolution_param {
		num_output: 256
		kernel_size: 5
		stride: 2
		weight_filler {
			type: "xavier"
		}
		bias_filler {
			type: "constant"
			value: 0
		}
	}
}

layer {
	name: "relu2"
	type: "ReLU"
	bottom: "conv2"
	top: "conv2"
}

layer {
	name: "pool2"
	type: "Pooling"
	bottom: "conv2"
	top: "pool2"
	pooling_param {
		kernel_size: 3
		stride: 2
		pool: MAX
	}
}

layer {
	name: "conv3"
	type: "Convolution"
	bottom: "pool2"
	top: "conv3"
	param { lr_mult: 1.0 }
	param { lr_mult: 2.0 }
	convolution_param {
		num_output: 512
		kernel_size: 3
		stride: 1
		weight_filler {
			type: "xavier"
		}
		bias_filler {
			type: "constant"
			value: 0
		}
	}
}
...
some other layers
...
layer {
	name:"fc8"
	bottom: "fc7"
	top: "fc8"
	type: "InnerProduct"
	inner_product_param {
		num_output: 512
		weight_filler {
			type: "xavier"
		}
		bias_filler {
			type: "constant"
			value: 0
		}
	}
}

layer {
    name: "sample_triplet"
    type: "SampleTriplet"
    bottom: "fc8"
    top: "triplet"
    sample_triplet_param {
      label_num: 8
      sample_num: 4
    }
}

layer {
    name: "triplet_loss"
    type: "TripletLoss"
    bottom: "fc8"
    bottom: "triplet"
    top: "triplet_loss"
    triplet_loss_param {
      margin: 0.25
    }
    loss_weight: 1
 }
`
训练一晚上后的结果，从开始到最后一直在0.75的小数点后4位波动
`
I0614 17:36:31.718358 28781 solver.cpp:244]     Train net output #32: triplet_loss = 0.749998 (* 1 = 0.749998 loss)
I0614 17:36:31.718369 28781 sgd_solver.cpp:106] Iteration 446000, lr = 1e-08
I0614 17:36:34.979749 28781 solver.cpp:228] Iteration 446020, loss = 0.75007
I0614 17:36:35.046409 28781 solver.cpp:244]     Train net output #0: label = 14537
I0614 17:36:35.046422 28781 solver.cpp:244]     Train net output #1: label = 14537
I0614 17:36:35.046438 28781 solver.cpp:244]     Train net output #2: label = 14537
I0614 17:36:35.046443 28781 solver.cpp:244]     Train net output #3: label = 14537
I0614 17:36:35.046448 28781 solver.cpp:244]     Train net output #4: label = 12832
I0614 17:36:35.046453 28781 solver.cpp:244]     Train net output #5: label = 12832
I0614 17:36:35.046458 28781 solver.cpp:244]     Train net output #6: label = 12832
I0614 17:36:35.046468 28781 solver.cpp:244]     Train net output #7: label = 12832
I0614 17:36:35.046478 28781 solver.cpp:244]     Train net output #8: label = 12280
I0614 17:36:35.046488 28781 solver.cpp:244]     Train net output #9: label = 12280
I0614 17:36:35.046499 28781 solver.cpp:244]     Train net output #10: label = 12280
I0614 17:36:35.046509 28781 solver.cpp:244]     Train net output #11: label = 12280
I0614 17:36:35.046519 28781 solver.cpp:244]     Train net output #12: label = 27357
I0614 17:36:35.046530 28781 solver.cpp:244]     Train net output #13: label = 27357
I0614 17:36:35.046540 28781 solver.cpp:244]     Train net output #14: label = 27357
I0614 17:36:35.046550 28781 solver.cpp:244]     Train net output #15: label = 27357
I0614 17:36:35.046561 28781 solver.cpp:244]     Train net output #16: label = 7992
I0614 17:36:35.046571 28781 solver.cpp:244]     Train net output #17: label = 7992
I0614 17:36:35.046581 28781 solver.cpp:244]     Train net output #18: label = 7992
I0614 17:36:35.046591 28781 solver.cpp:244]     Train net output #19: label = 7992
I0614 17:36:35.046602 28781 solver.cpp:244]     Train net output #20: label = 9613
I0614 17:36:35.046613 28781 solver.cpp:244]     Train net output #21: label = 9613
I0614 17:36:35.046623 28781 solver.cpp:244]     Train net output #22: label = 9613
I0614 17:36:35.046634 28781 solver.cpp:244]     Train net output #23: label = 9613
I0614 17:36:35.046645 28781 solver.cpp:244]     Train net output #24: label = 19435
I0614 17:36:35.046655 28781 solver.cpp:244]     Train net output #25: label = 19435
I0614 17:36:35.046665 28781 solver.cpp:244]     Train net output #26: label = 19435
I0614 17:36:35.046672 28781 solver.cpp:244]     Train net output #27: label = 19435
I0614 17:36:35.046682 28781 solver.cpp:244]     Train net output #28: label = 219
I0614 17:36:35.046694 28781 solver.cpp:244]     Train net output #29: label = 219
I0614 17:36:35.046703 28781 solver.cpp:244]     Train net output #30: label = 219
I0614 17:36:35.046715 28781 solver.cpp:244]     Train net output #31: label = 219
I0614 17:36:35.046730 28781 solver.cpp:244]     Train net output #32: triplet_loss = 0.75007 (* 1 = 0.75007 loss)
I0614 17:36:35.046741 28781 sgd_solver.cpp:106] Iteration 446020, lr = 1e-08
I0614 17:36:38.170933 28781 solver.cpp:228] Iteration 446040, loss = 0.750116
I0614 17:36:38.238198 28781 solver.cpp:244]     Train net output #0: label = 384
I0614 17:36:38.238209 28781 solver.cpp:244]     Train net output #1: label = 384
I0614 17:36:38.238226 28781 solver.cpp:244]     Train net output #2: label = 384
I0614 17:36:38.238231 28781 solver.cpp:244]     Train net output #3: label = 384
I0614 17:36:38.238235 28781 solver.cpp:244]     Train net output #4: label = 11448
I0614 17:36:38.238240 28781 solver.cpp:244]     Train net output #5: label = 11448
I0614 17:36:38.238246 28781 solver.cpp:244]     Train net output #6: label = 11448
I0614 17:36:38.238256 28781 solver.cpp:244]     Train net output #7: label = 11448
I0614 17:36:38.238266 28781 solver.cpp:244]     Train net output #8: label = 9058
I0614 17:36:38.238276 28781 solver.cpp:244]     Train net output #9: label = 9058
I0614 17:36:38.238287 28781 solver.cpp:244]     Train net output #10: label = 9058
I0614 17:36:38.238297 28781 solver.cpp:244]     Train net output #11: label = 9058
I0614 17:36:38.238308 28781 solver.cpp:244]     Train net output #12: label = 26239
I0614 17:36:38.238319 28781 solver.cpp:244]     Train net output #13: label = 26239
I0614 17:36:38.238348 28781 solver.cpp:244]     Train net output #14: label = 26239
I0614 17:36:38.238359 28781 solver.cpp:244]     Train net output #15: label = 26239
I0614 17:36:38.238369 28781 solver.cpp:244]     Train net output #16: label = 15103
I0614 17:36:38.238380 28781 solver.cpp:244]     Train net output #17: label = 15103
I0614 17:36:38.238390 28781 solver.cpp:244]     Train net output #18: label = 15103
I0614 17:36:38.238401 28781 solver.cpp:244]     Train net output #19: label = 15103
I0614 17:36:38.238411 28781 solver.cpp:244]     Train net output #20: label = 4531
I0614 17:36:38.238421 28781 solver.cpp:244]     Train net output #21: label = 4531
I0614 17:36:38.238432 28781 solver.cpp:244]     Train net output #22: label = 4531
I0614 17:36:38.238442 28781 solver.cpp:244]     Train net output #23: label = 4531
I0614 17:36:38.238452 28781 solver.cpp:244]     Train net output #24: label = 8527
I0614 17:36:38.238463 28781 solver.cpp:244]     Train net output #25: label = 8527
I0614 17:36:38.238473 28781 solver.cpp:244]     Train net output #26: label = 8527
I0614 17:36:38.238484 28781 solver.cpp:244]     Train net output #27: label = 8527
I0614 17:36:38.238494 28781 solver.cpp:244]     Train net output #28: label = 17737
I0614 17:36:38.238505 28781 solver.cpp:244]     Train net output #29: label = 17737
I0614 17:36:38.238515 28781 solver.cpp:244]     Train net output #30: label = 17737
I0614 17:36:38.238525 28781 solver.cpp:244]     Train net output #31: label = 17737
I0614 17:36:38.238540 28781 solver.cpp:244]     Train net output #32: triplet_loss = 0.750116 (* 1 = 0.750116 loss)
I0614 17:36:38.238554 28781 sgd_solver.cpp:106] Iteration 446040, lr = 1e-08

`

训练数据的话有约3k个类。不过我们的图像比较凌乱，大小不一，没什么预处理，不知道这会不会有影响。
谢谢！

**Comment by tyandzx** [comment]
sorry，不知道为什么延迟了很久才看到你的消息，非常抱歉。这个仓库已经许久不更新了，所以可能回复的比较迟。
针对你的问题，
1）我们采用的是利用向量内积来计算两个样本的相似度。具体的公式很简单，针对向量A和B，那么内积的计算方式为:sum(A.*B)/(norm(A)*norm(B))。
2）loss不下降的准确原因我也不是很确定，我提供几个思路：a.一个是margin我觉得你设的可能过大，因为我们采用的是余弦相似度，你可以通过acos(margin)来计算一下对应的角度，这个角度是你希望类内样本之间的夹角减去类间样本之间的夹角。我觉得0.25通过上面的计算，那么这个夹角应该比较大了，况且你有3k个类，很难做到每两个类间都能保持这么大的夹角。b.另一个是我个人认为triplet loss加在网络的倒数第二层比加在网络输出的概率上更合适，原因是我们更倾向于使学习到的特征空间具有聚类的性质，而不是网络的输出。也就是说，triplet loss我比较倾向于加在fc7上，而不是fc8上。
